### PR TITLE
[llvm][NVPTX] Fix quadratic runtime in ProxyRegErasure

### PR DESCRIPTION
**Pull request by Mogball**
This pass performs RAUW by walking the machine function for each RAUW operation. For large functions, this runtime in this pass starts to blow up. Linearize the pass by batching the RAUW ops at once.

### CODE REVIEW

File: llvm/lib/Target/NVPTX/NVPTXProxyRegErasure.cpp
Comment (justinfargnoli) [inline llvm/lib/Target/NVPTX/NVPTXProxyRegErasure.cpp:101]: Ignorable nit: prefer early return.

**Review by justinfargnoli — COMMENTED** [review]
Thanks for the fix, @Mogball!

Is it not possible to write compile-time tests in LLVM?

**Comment by Mogball** [comment]
> Thanks for the fix, @Mogball!
> 
> Is it not possible to write compile-time tests in LLVM?

I'm not aware of any infrastructure or testing regimes for doing something like that. Accurate benchmarking is less trivial than running unit tests :P

File: llvm/lib/Target/NVPTX/NVPTXProxyRegErasure.cpp
Comment (Mogball) [reply llvm/lib/Target/NVPTX/NVPTXProxyRegErasure.cpp]: done. Thanks!

**Comment by Mogball** [comment]
NVPTX backend compile times are top-of-mind for me right now. So if there are any problems, you'll definitely hear about them from me!

**Review by justinfargnoli — APPROVED** [review]
LGTM!

Please wait for another community member to approve the MR before merging.

**Review by lattner — APPROVED** [review]
Looks like a straight-forward approach, thanks!

**Review by Artem-B — APPROVED** [review]
Nice. Thank you for the fix.

**Comment by joker-eph** [comment]
There is a compile-time infrastructure using clang and the test-suite, so we should be able to have Cuda example these as well I guess?

**Comment by gonzalobg** [comment]
Nice @Mogball , thank you !

**Comment by Groverkss** [comment]
Hi! This pr seems to break nvptx compilation. I was able to get a reduced test case:

```
llc test.ll
```

test.ll: https://gist.github.com/Groverkss/d6205367d8bdaff9b848fc14cc9a43a7

**Comment by Groverkss** [comment]
> @Groverkss -- what happens when we compile the reproducer? Does it generate invalid output or is it a compiler crasher?

Oops, sorry, I should have included the failure message. The compiler crashes on an assert (I have a RelWithDebInfo build):

https://gist.github.com/Groverkss/1160f651b063df57303b4f8112435b27

**Comment by Mogball** [comment]
Hey thanks for the repro. I'm looking right now. Please feel free to revert in the mean time.

**Comment by Mogball** [comment]
@Groverkss I'm not able to reproduce this. Do you have a commit hash I can try? Also, you're on the cmake relwithdebinfo config right?

**Comment by Groverkss** [comment]
Thanks for looking into this. I'm trying this on 08e5a1de8227512d4774a534b91cb2353cef6284 .

I checked again, and the configuration I'm on is Release with assertions enabled.

**Comment by Mogball** [comment]
Got it. I found the bug. The issue is something like

```
%1 = mov %0
%2 = mov %1

use %0
use %0
use %1
```

I have a fix and I'm just trying to figure out how to write an MIR test!

**Comment by Groverkss** [comment]
Great! Feel free to fix it forward, I'm not blocked on this and not in need of a revert.

**Comment by Mogball** [comment]
Fixed in https://github.com/llvm/llvm-project/pull/105871

**Pull request merged**
